File: recipes/subms-timer-wheel/java/src/main/java/com/submillisecond/recipes/timer/SampleApp.java
package com.submillisecond.recipes.timer;

import com.submillisecond.recipes.timer.features.ConcurrentTimerWheel;
import com.submillisecond.recipes.timer.features.CronSchedule;
import com.submillisecond.recipes.timer.features.CronScheduler;
import com.submillisecond.recipes.timer.features.DeadlineScheduler;
import com.submillisecond.recipes.timer.features.HierarchicalTimerWheel;
import com.submillisecond.recipes.timer.features.MeteredTimerWheel;
import com.submillisecond.recipes.timer.features.TestClock;
import com.submillisecond.recipes.timer.features.TimerMetrics;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Sample app: an order-lifecycle supervisor on a matching engine, built on
 * {@code subms-timer-wheel}. Run:
 * {@code mvn -q compile exec:java -Dexec.mainClass=com.submillisecond.recipes.timer.SampleApp}
 *
 * <p>Every resting order arms an expiry timer sized to its time-in-force. A
 * fill cancels it, an amend reschedules it, and end-of-session drains what is
 * left. Session time is a tick counter and the clock-driven sections use an
 * injected clock, so the printed output is identical on every run.
 *
 * <ul>
 *   <li>base               - the TIF supervisor: arm, cancel on fill, amend, drain
 *   <li>hierarchical       - good-til-date orders whose horizons span seconds to a session
 *   <li>concurrent         - quote-timeout timers armed from many market-data threads
 *   <li>deadline-scheduler - FIX session idle timeout, bumped by inbound traffic
 *   <li>cron               - a recurring mark-to-market risk snapshot
 *   <li>metrics            - the supervisor reporting its own cadence
 * </ul>
 */
public final class SampleApp {

    /** What the matching engine hands the supervisor, at a given session second. */
    private record Event(int at, String kind, String order, int tif) { }

    public static void main(String[] args) throws InterruptedException {
        tifSupervisor();
        hierarchicalGtd();
        concurrentQuoteTimeouts();
        deadlineSessionIdle();
        cronRiskSnapshot();
        meteredExpiryWheel();
    }

    /**
     * Base API. One tick is one second of session time. The supervisor holds an
     * order id to timer id map because the wheel hands back a timer id, and the
     * engine only ever speaks order ids.
     */
    static void tifSupervisor() {
        System.out.println("== base: order time-in-force supervisor ==");

        List<Event> tape = List.of(
            new Event(0, "rest", "ORD-A", 3),
            new Event(0, "rest", "ORD-B", 5),
            new Event(0, "rest", "ORD-C", 9),
            new Event(0, "rest", "ORD-D", 12),
            new Event(2, "fill", "ORD-B", 0),
            new Event(4, "amend", "ORD-C", 6));

        TimerWheel<String> expiries = new TimerWheel<>(256);
        Map<String, Long> timerOf = new HashMap<>();

        int sessionSecs = 11;
        for (int second = 0; second <= sessionSecs; second++) {
            for (Event ev : tape) {
                if (ev.at() != second) continue;
                switch (ev.kind()) {
                    case "rest" -> {
                        timerOf.put(ev.order(), expiries.schedule(ev.tif(), ev.order()));
                        System.out.println("  t=" + second + "s rest " + ev.order() + " tif=" + ev.tif() + "s");
                    }
                    case "fill" -> {
                        expiries.cancel(timerOf.get(ev.order()));
                        System.out.println("  t=" + second + "s fill " + ev.order() + " -> expiry cancelled");
                    }
                    default -> {
                        expiries.reschedule(timerOf.get(ev.order()), ev.tif());
                        System.out.println("  t=" + second + "s amend " + ev.order()
                            + " tif -> " + ev.tif() + "s from now");
                    }
                }
            }
            if (second == sessionSecs) break;
            for (String ord : expiries.tick()) {
                System.out.println("  t=" + (second + 1) + "s expire " + ord);
            }
        }

        List<String> unfilled = expiries.drain();
        System.out.println("  session close: " + unfilled.size() + " orders still resting " + unfilled);
        System.out.println("  pending after drain: " + expiries.pending());

        if (!unfilled.equals(List.of("ORD-D"))) {
            throw new AssertionError("only the 12s TIF outlives the session, got " + unfilled);
        }
        if (expiries.pending() != 0) throw new AssertionError("every timer retired");
    }

    /**
     * hierarchical: good-til-date orders expire anywhere from a few seconds to a
     * full session out. The hierarchical wheel holds far-out orders on a coarse
     * level and cascades them down as their deadline approaches, from a fixed
     * 192-bucket footprint.
     */
    static void hierarchicalGtd() {
        System.out.println("\n== hierarchical: good-til-date across horizons ==");
        HierarchicalTimerWheel<String> gtd = new HierarchicalTimerWheel<>();
        gtd.schedule(30, "GTD-near");            // intraday, lands on the fine wheel
        long far = gtd.schedule(5000, "GTD-far"); // deep on the coarse wheel
        System.out.println("  armed 2 GTD orders, " + gtd.pending() + " pending");

        // The desk pulls the far order in to the close of the current session.
        gtd.reschedule(far, 300);
        System.out.println("  GTD-far pulled in to t=300");

        Long nearAt = null, farAt = null;
        for (int t = 1; t <= 300; t++) {
            for (String id : gtd.tick()) {
                if (id.equals("GTD-near")) nearAt = (long) t;
                else if (id.equals("GTD-far")) farAt = (long) t;
            }
        }
        System.out.println("  near fired at t=" + nearAt + ", far fired at t=" + farAt);
        System.out.println("  cascade events: " + gtd.cascades());
        if (nearAt == null || nearAt != 30L) throw new AssertionError("near GTD fires on its deadline");
        if (farAt == null || farAt != 300L) throw new AssertionError("the rescheduled GTD fires on its new deadline");
        if (gtd.cascades() < 1) throw new AssertionError("the far order cascaded down a level");
        if (gtd.pending() != 0) throw new AssertionError("every GTD retired");
    }

    /** concurrent: quote-timeout timers armed from many market-data threads. */
    static void concurrentQuoteTimeouts() throws InterruptedException {
        System.out.println("\n== concurrent: quote timeouts from many feeds ==");
        ConcurrentTimerWheel<Integer> wheel = new ConcurrentTimerWheel<>(256);
        int feeds = 4, perFeed = 50;
        List<Thread> threads = new ArrayList<>();
        for (int feed = 0; feed < feeds; feed++) {
            final int f = feed;
            Thread t = new Thread(() -> {
                for (int i = 0; i < perFeed; i++) wheel.schedule(1 + (i % 8), f * 1000 + i);
            });
            threads.add(t);
            t.start();
        }
        for (Thread t : threads) t.join();
        System.out.println("  " + wheel.pending() + " quote timeouts armed across " + feeds + " feeds");

        int fired = wheel.advance(16).size();
        System.out.println("  " + feeds + " feeds x " + perFeed + " quotes -> " + fired + " timeouts fired");
        if (fired != feeds * perFeed) throw new AssertionError("every armed timeout fires once");
        if (!wheel.isEmpty()) throw new AssertionError("the wheel drained itself");
    }

    /**
     * deadline-scheduler: a FIX session must see inbound traffic inside its idle
     * window or be torn down. One timer per session, bumped on every message
     * rather than cancelled and re-armed. A hand-stepped clock keeps the demo
     * deterministic instead of sleeping.
     */
    static void deadlineSessionIdle() {
        System.out.println("\n== deadline-scheduler: FIX session idle timeout ==");
        Duration idle = Duration.ofMillis(30);
        TestClock clock = new TestClock();
        DeadlineScheduler<String> sched = new DeadlineScheduler<>(256, clock, Duration.ofMillis(1));

        long session = sched.scheduleAfter(idle, "SESSION-1");
        long elapsed = 0;
        for (int gap : new int[] {10, 15}) {
            clock.advance(Duration.ofMillis(gap));
            elapsed += gap;
            if (!sched.poll().isEmpty()) throw new AssertionError("traffic keeps the session alive");
            sched.rescheduleAfter(session, idle);
            System.out.println("  inbound msg at +" + elapsed + "ms, idle deadline now +" + (elapsed + 30) + "ms");
        }

        clock.advance(Duration.ofMillis(30));
        List<String> dead = sched.poll();
        System.out.println("  no traffic for " + idle.toMillis() + "ms -> " + dead);
        if (!dead.equals(List.of("SESSION-1"))) throw new AssertionError("the idle timeout fires");
    }

    /** cron: a mark-to-market risk snapshot every 5 minutes, re-arming each fire. */
    static void cronRiskSnapshot() {
        System.out.println("\n== cron: mark-to-market every 5 minutes ==");
        CronSchedule schedule = CronSchedule.parse("*/5 * * * *");
        long start = 1_704_067_201L; // 2024-01-01 00:00:01 UTC
        CronScheduler scheduler = new CronScheduler(schedule, start);

        long first = scheduler.nextFire(start);
        scheduler.recordFire(first);
        long second = scheduler.nextFire(first);
        System.out.println("  first snapshot at epoch " + first + ", next at " + second);
        if (first != 1_704_067_500L) throw new AssertionError("first fire lands on the 5-minute grid");
        if (second != first + 300) throw new AssertionError("re-arms exactly 5 minutes later");
    }

    /**
     * metrics: the supervisor reports its own cadence. A drained timer is
     * counted apart from a fired one, so a session close does not read as a
     * burst of expiries.
     */
    static void meteredExpiryWheel() {
        System.out.println("\n== metrics: self-reporting expiry counters ==");
        MeteredTimerWheel<String> wheel = new MeteredTimerWheel<>(64);
        wheel.schedule(2, "ORD-A");
        long b = wheel.schedule(2, "ORD-B");
        long c = wheel.schedule(2, "ORD-C");
        wheel.cancel(b);
        wheel.reschedule(c, 20);

        int fired = wheel.advance(3).size();
        List<String> left = wheel.drain();
        TimerMetrics m = wheel.metrics();
        System.out.println("  scheduled=" + m.scheduled + " fired=" + m.fired
            + " cancelled=" + m.cancelled + " rescheduled=" + m.rescheduled
            + " drained=" + m.drained + " ticks=" + m.ticks);

        if (m.scheduled != 3 || m.cancelled != 1 || m.rescheduled != 1 || m.fired != 1
            || m.drained != 1 || m.ticks != 3 || fired != 1 || !left.equals(List.of("ORD-C"))) {
            throw new AssertionError("counters must reflect one fire, one cancel, one reschedule");
        }
    }
}

File: recipes/subms-timer-wheel/rust/examples/sample_app.rs
//! Sample app: an order-lifecycle supervisor on a matching engine, built on
//! `subms-timer-wheel`. Run the base with `cargo run --example sample_app`;
//! add `--features full` to see each opt-in section light up.
//!
//! Every resting order arms an expiry timer sized to its time-in-force. A
//! fill cancels it, an amend reschedules it, and end-of-session drains what
//! is left. Session time is a tick counter and the clock-driven sections use
//! an injected clock, so the printed output is identical on every run.
//!
//! * base               - the TIF supervisor: arm, cancel on fill, amend, drain
//! * hierarchical       - good-til-date orders whose horizons span seconds to a session
//! * concurrent         - quote-timeout timers armed from many market-data threads
//! * deadline-scheduler - FIX session idle timeout, bumped by inbound traffic
//! * cron               - a recurring mark-to-market risk snapshot
//! * metrics            - the supervisor reporting its own cadence

use subms_timer_wheel::TimerWheel;

fn main() {
    tif_supervisor();

    #[cfg(feature = "hierarchical")]
    hierarchical_gtd();

    #[cfg(feature = "concurrent")]
    concurrent_quote_timeouts();

    #[cfg(feature = "deadline-scheduler")]
    deadline_session_idle();

    #[cfg(feature = "cron")]
    cron_risk_snapshot();

    #[cfg(feature = "metrics")]
    metered_expiry_wheel();
}

/// What the matching engine hands the supervisor, at a given session second.
enum Event {
    /// A new resting order with a time-in-force in seconds.
    Rest(&'static str, usize),
    /// Fully filled: its expiry must not fire.
    Fill(&'static str),
    /// Amended to a longer time-in-force, measured from now.
    Amend(&'static str, usize),
}

/// Base API. One tick is one second of session time. The supervisor holds an
/// order id to timer id map because the wheel hands back a timer id, and the
/// engine only ever speaks order ids.
fn tif_supervisor() {
    println!("== base: order time-in-force supervisor ==");

    let tape = [
        (0usize, Event::Rest("ORD-A", 3)),
        (0, Event::Rest("ORD-B", 5)),
        (0, Event::Rest("ORD-C", 9)),
        (0, Event::Rest("ORD-D", 12)),
        (2, Event::Fill("ORD-B")),
        (4, Event::Amend("ORD-C", 6)),
    ];

    let mut expiries: TimerWheel<&'static str> = TimerWheel::new(256);
    let mut timer_of: Vec<(&'static str, u64)> = Vec::new();
    let lookup = |map: &Vec<(&'static str, u64)>, ord: &str| {
        map.iter().find(|(o, _)| *o == ord).map(|(_, t)| *t)
    };

    let session_secs = 11;
    for second in 0..=session_secs {
        for (at, ev) in tape.iter() {
            if *at != second {
                continue;
            }
            match ev {
                Event::Rest(ord, tif) => {
                    let id = expiries.schedule(*tif, ord);
                    timer_of.push((ord, id));
                    println!("  t={second}s rest {ord} tif={tif}s");
                }
                Event::Fill(ord) => {
                    let id = lookup(&timer_of, ord).expect("a resting order");
                    expiries.cancel(id);
                    println!("  t={second}s fill {ord} -> expiry cancelled");
                }
                Event::Amend(ord, tif) => {
                    let id = lookup(&timer_of, ord).expect("a resting order");
                    expiries.reschedule(id, *tif);
                    println!("  t={second}s amend {ord} tif -> {tif}s from now");
                }
            }
        }
        if second == session_secs {
            break;
        }
        for ord in expiries.tick() {
            println!("  t={}s expire {ord}", second + 1);
        }
    }

    let unfilled = expiries.drain();
    println!(
        "  session close: {} orders still resting {:?}",
        unfilled.len(),
        unfilled
    );
    println!("  pending after drain: {}", expiries.pending());

    assert_eq!(
        unfilled,
        vec!["ORD-D"],
        "only the 12s TIF outlives the session"
    );
    assert_eq!(expiries.pending(), 0);
}

/// `hierarchical` feature: good-til-date orders expire anywhere from a few
/// seconds to a full session out. A single flat wheel would need a slot per
/// tick of the longest horizon; the hierarchical wheel holds far-out orders
/// on a coarse level and cascades them down as their deadline approaches,
/// from a fixed 192-bucket footprint.
#[cfg(feature = "hierarchical")]
fn hierarchical_gtd() {
    use subms_timer_wheel::HierarchicalTimerWheel;
    println!("\n== hierarchical: good-til-date across horizons ==");
    let mut gtd: HierarchicalTimerWheel<&'static str> = HierarchicalTimerWheel::new();

    gtd.schedule(30, "GTD-near"); // intraday, lands on the fine wheel
    let far = gtd.schedule(5000, "GTD-far"); // deep on the coarse wheel
    println!("  armed 2 GTD orders, {} pending", gtd.pending());

    // The desk pulls the far order in to the close of the current session.
    gtd.reschedule(far, 300);
    println!("  GTD-far pulled in to t=300");

    let mut near_at = None;
    let mut far_at = None;
    for t in 1..=300 {
        for id in gtd.tick() {
            match id {
                "GTD-near" => near_at = Some(t),
                "GTD-far" => far_at = Some(t),
                _ => {}
            }
        }
    }
    println!(
        "  near fired at t={:?}, far fired at t={:?}",
        near_at, far_at
    );
    println!("  cascade events: {}", gtd.cascades());

    assert_eq!(near_at, Some(30), "near GTD fires on its deadline");
    assert_eq!(
        far_at,
        Some(300),
        "the rescheduled GTD fires on its new deadline"
    );
    assert!(gtd.cascades() >= 1, "the far order cascaded down a level");
    assert_eq!(gtd.pending(), 0);
}

/// `concurrent` feature: several market-data threads each arm quote-timeout
/// timers against one shared wheel. `Clone` shares the handle; every op
/// serializes on a short mutex. A single ticker thread then drains expiries.
#[cfg(feature = "concurrent")]
fn concurrent_quote_timeouts() {
    use std::thread;
    use subms_timer_wheel::ConcurrentTimerWheel;
    println!("\n== concurrent: quote timeouts from many feeds ==");
    let wheel: ConcurrentTimerWheel<usize> = ConcurrentTimerWheel::new(256);

    let feeds = 4;
    let per_feed = 50;
    let mut handles = Vec::new();
    for feed in 0..feeds {
        let wheel = wheel.clone();
        handles.push(thread::spawn(move || {
            for i in 0..per_feed {
                wheel.schedule(1 + (i % 8), feed * 1000 + i);
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    println!(
        "  {} quote timeouts armed across {feeds} feeds",
        wheel.pending()
    );

    let fired = wheel.advance(16).len();
    println!("  {feeds} feeds x {per_feed} quotes -> {fired} timeouts fired");
    assert_eq!(
        fired,
        feeds * per_feed,
        "every armed timeout fired exactly once"
    );
    assert!(wheel.is_empty());
}

/// `deadline-scheduler` feature: a FIX session must see inbound traffic
/// inside its idle window or be torn down. One timer per session, bumped on
/// every message rather than cancelled and re-armed. Callers think in
/// instants; the layer maps them to ticks through the injected clock, and
/// `poll()` fires the catch-up batch. A hand-stepped clock keeps the demo
/// deterministic instead of sleeping.
#[cfg(feature = "deadline-scheduler")]
fn deadline_session_idle() {
    use std::time::Duration;
    use subms_timer_wheel::{DeadlineScheduler, TestClock};
    println!("\n== deadline-scheduler: FIX session idle timeout ==");

    let idle = Duration::from_millis(30);
    let mut sched: DeadlineScheduler<&'static str, TestClock> =
        DeadlineScheduler::new(256, TestClock::new(), Duration::from_millis(1));

    let session = sched.schedule_after(idle, "SESSION-1");
    let mut elapsed = 0u64;
    for gap in [10u64, 15] {
        sched.clock().advance(Duration::from_millis(gap));
        elapsed += gap;
        assert!(sched.poll().is_empty(), "traffic keeps the session alive");
        sched.reschedule_after(session, idle);
        println!(
            "  inbound msg at +{elapsed}ms, idle deadline now +{}ms",
            elapsed + 30
        );
    }
    // Then the counterparty goes quiet.
    sched.clock().advance(Duration::from_millis(30));
    let dead = sched.poll();
    println!("  no traffic for {}ms -> {:?}", idle.as_millis(), dead);
    assert_eq!(dead, vec!["SESSION-1"], "the idle timeout fires");
}

/// `cron` feature: a recurring risk snapshot on a wall-clock cadence. The
/// scheduler parses the 5-field expression once, then re-arms the next
/// matching second each time the current one fires.
#[cfg(feature = "cron")]
fn cron_risk_snapshot() {
    use subms_timer_wheel::{CronSchedule, CronScheduler};
    println!("\n== cron: mark-to-market every 5 minutes ==");
    let schedule = CronSchedule::parse("*/5 * * * *").expect("valid cron");

    // 2024-01-01 00:00:01 UTC. Next */5 boundary is 00:05:00.
    let start = 1_704_067_201;
    let mut scheduler = CronScheduler::new(schedule, start);

    let first = scheduler.next_fire(start).expect("a next fire exists");
    scheduler.record_fire(first);
    let second = scheduler.next_fire(first).expect("a next fire exists");
    println!("  first snapshot at epoch {first}, next at {second}");

    assert_eq!(
        first, 1_704_067_500,
        "first fire lands on the 5-minute grid"
    );
    assert_eq!(second, first + 300, "re-arms exactly 5 minutes later");
}

/// `metrics` feature: the supervisor reports its own cadence. The counters
/// are plain fields (the wheel is single-threaded), read through a snapshot.
/// A drained timer is counted apart from a fired one, so a session close does
/// not read as a burst of expiries.
#[cfg(feature = "metrics")]
fn metered_expiry_wheel() {
    use subms_timer_wheel::MeteredTimerWheel;
    println!("\n== metrics: self-reporting expiry counters ==");
    let mut wheel: MeteredTimerWheel<&'static str> = MeteredTimerWheel::new(64);

    let a = wheel.schedule(2, "ORD-A");
    let b = wheel.schedule(2, "ORD-B");
    let c = wheel.schedule(2, "ORD-C");
    wheel.cancel(b);
    wheel.reschedule(c, 20);
    let _ = a;

    let fired = wheel.advance(3).len();
    let left = wheel.drain();
    let m = wheel.metrics();
    println!(
        "  scheduled={} fired={} cancelled={} rescheduled={} drained={} ticks={}",
        m.scheduled, m.fired, m.cancelled, m.rescheduled, m.drained, m.ticks
    );

    assert_eq!(m.scheduled, 3);
    assert_eq!(m.cancelled, 1);
    assert_eq!(m.rescheduled, 1);
    assert_eq!(fired, 1, "only the untouched order fired");
    assert_eq!(m.fired, 1);
    assert_eq!(left, vec!["ORD-C"]);
    assert_eq!(m.drained, 1);
    assert_eq!(m.ticks, 3);
}
